Using with operators (#12362)
ethereum/solidity: docs/contracts/using-for.rst

Using For
*********

The directive ``using A for B;`` can be used to attach functions (``A``)

> as operators to user-defined type or as member functions to any type (``B``).

Reviewer: ```suggestion
as operators to user-defined types or as member functions to any type (``B``).
```

Reviewer 2: fixed

The member functions receive the object they are called on as their first
parameter (like the ``self`` variable in Python). The operator functions

> receives operands as parameters.

Reviewer: ```suggestion
receive operands as parameters.
```

Reviewer 2: fixed

It is valid either at file level or inside a contract,
at contract level.

The first part, ``A``, can be one of:

- a list of file-level or library functions (``using {f, g, h, L.t} for uint;``) -
  only those functions will be attached to the type as member functions,
- the name of a library (``using L for uint;``) -
  all functions (both public and internal ones) of the library are attached to the type
  as member functions,
- a list of assignments of functions to operators (``using {f as +, g as -} for T;``) -
  the functions will be attached to the type (``T``) as operators.

At file level, the second part, ``B``, has to be an explicit type (without data location specifier).
To attach functions as member functions inside contracts, you can also use ``using L for *;``, which

> has the effect that all functions of the library ``L`` are attached to *all* types.

Reviewer: I think the old wording was clearer. Now it sounds as if you were attaching member functions to contracts.

Reviewer 2: OK. The change reverted.

Using For
*********

The directive ``using A for B;`` can be used to bind functions (``A``)

> as operators to user-defined types or as member functions to any type (``B``).

Reviewer: ```suggestion
as operators to user-defined value types and structs or as member functions to any type (``B``).
```

Reviewer 2: done

The member functions receive the object they are called on as their first
parameter (like the ``self`` variable in Python). The operator functions
receive operands as parameters.

It is valid either at file level or inside a contract,
at contract level.

The first part, ``A``, can be one of:

- a list of file-level or library functions (``using {f, g, h, L.t} for uint;``) -
  only those functions will be bound to the type as member functions,
- the name of a library (``using L for uint;``) -
  all functions (both public and internal ones) of the library are bound to the type
  as member functions,
- a list of assignments of functions to operators (``using {f as +, g as -} for T;``) -

>   the functions will be bound to the type (``T``) as operators.

Reviewer: I think we should document which operators can be customized. The downside we'll have to keep that up to date but on the other hand I think users should be able to learn that from the docs and not have to dig in the source.

Also, which of them can be unary (it's probably not obvious that e.g. `+` can't).

Reviewer 2: Documentation updated with lists of unary and binary operators that can be customized.

>   the functions will be bound to the type (``T``) as operators.

Reviewer: We should mention that these also must be file-level or internal library functions. And the point about `using {f} for T` does not mention it but it also requires the library function to be internal.

Reviewer: We should also say that in all these cases these functions cannot be overloaded. If you define them for more than one type, you'll get an error.

Reviewer 2: Information about requirements on functions added. However, I am not sure if an extra information about what is not allowed here is a good idea. The list of forbidden things might be pretty long.

Reviewer: I think it won't be that bad if we don't go into tiny details. I think that overloading is important enough to be mentioned.

>   the functions will be bound to the type (``T``) as operators.

Reviewer: We should mention that these functions are **not** at the same time available as bound functions. They have to be bound separately.

Reviewer 2: Again, I am not a huge fan of describing things that are not happening.

Reviewer: ok, if you don't like it I guess we can skip that one. I'd add it but this is just a minor detail.

Currently these docs look more like a reference than a user-friendly guide and a reference should have such details. I think at some point we should decide which way we want to go. That's a discussion for another day though.

first parameter of each of these functions. This check is
performed even if none of these functions are called.

> If you define a user type operator (``using {f as +} for T``), then

Reviewer: ```suggestion
If you define an operator for a user-defined type (``using {f as +} for T``), then
```

Reviewer 2: updated

> the type (``T``), types of function parameters and a type of function return value

Reviewer: ```suggestion
the type (``T``), types of function parameters and the type of the function return value
```

Reviewer 2: fixed

> has to be the same. The type (``T``) does not include data location.

Reviewer: ```suggestion
the type (``T``), types of function parameters and the type of function return value
have to be the same. The type (``T``) does not include data location.
```

Reviewer 2: updated

> have to be the same. The type (``T``) does not include data location.

Reviewer: This should mention that this does not apply to comparison operators. They must return `bool`.

Reviewer 2: done

When the directive is used at file level and applied to a
user-defined type which was defined at file level in the same file,
the word ``global`` can be added at the end. This will have the

> effect that the functions are bound to the type everywhere

Reviewer: ```suggestion
effect that the functions and operators are bound to the type everywhere
```

Reviewer 2: done

``self`` variable. The only situation where no copy will be performed
is when storage reference variables are used or when internal library
functions are called.

> Another example shows how to define user type operator:

Reviewer: ```suggestion
Another example shows how to define a custom operator for a user-defined type:
```

Reviewer 2: done

.. code-block:: solidity

    // SPDX-License-Identifier: GPL-3.0

>     pragma solidity >=0.8.17 <0.9.0;

Reviewer: ```suggestion
    pragma solidity ^0.8.17;
```

Reviewer 2: done

    using {add as +} for Point;

    struct Point {
        uint x;
        uint y;
    }

>     function add(Point memory _a, Point memory _b) pure returns (Point memory result) {

Reviewer: We don't really use the underscore convention in Soldiity code (#11764):
```suggestion
    function add(Point memory a, Point memory b) pure returns (Point memory result) {
```

Reviewer 2: code updated

        result.x = _a.x + _b.x;
        result.y = _a.y + _b.y;
    }

    contract C {

>         function test() pure public {

Reviewer: Things like `C` or `test` are fine in test cases but I think that for the examples we have in docs we should try to come up with something at least a little more meaningful. For example you could have the contract accept two points and return the distance between them or something like that.

Reviewer: And actually, since the primary motivation behind operators is to allow implementing fixed-point types, an example related to that would be even better.

Reviewer 2: done

            Point memory p1 = Point({x:3, y:4});

>             Point memory p2 = Point({x:7, y:16});

Reviewer: ```suggestion
            Point memory p1 = Point({x: 3, y: 4});
            Point memory p2 = Point({x: 7, y: 16});
```

Reviewer 2: done

``self`` variable. The only situation where no copy will be performed
is when storage reference variables are used or when internal library
functions are called.

Another example shows how to define a custom operator for a user-defined type:

.. code-block:: solidity

    // SPDX-License-Identifier: GPL-3.0
    pragma solidity ^0.8.18;

    type UFixed16x2 is uint16;

    using {
        add as +,
        div as /

>     } for UFixed16x2;

Reviewer: Maybe it would be good to have `-` instead of `+` here and show that you can attach two different functions to it to get binary and unary version working?

Reviewer 2: I would like to keep it as simple as possible to not overwhelm a user with a sophisticated example that contains many ambushed language features.

Reviewer: The reason I'm suggesting that is that I would consider redefining an unary operator to be a common thing, not something arcane. On the other hand, it might not be obvious to people that listing an operator twice is even allowed in the first place, much less expected. As a user I'd expect the documentation to show these things to me.

If not in the example, I think we should mention it in the docs above. I just thought that weaving it into the example would be a nicer way to show such a detail it without making the text more complex.

Reviewer 2: I've updated the documentation so now it says that it is possible to have the same operator defined both unary and binary in the list of single using directive.

    uint32 constant SCALE = 100;

    function add(UFixed16x2 a, UFixed16x2 b) pure returns (UFixed16x2) {
        return UFixed16x2.wrap(UFixed16x2.unwrap(a) + UFixed16x2.unwrap(b));
    }

    function div(UFixed16x2 a, UFixed16x2 b) pure returns (UFixed16x2) {
        uint32 a32 = UFixed16x2.unwrap(a);
        uint32 b32 = UFixed16x2.unwrap(b);

>         return UFixed16x2.wrap(uint16(a32 * SCALE / b32));

Reviewer: The `uint16` conversion will truncate when `a` is large and `b` is small. Consider for example `655.35 / 0.01`, which produces a result that cannot be represented in the type. Please add an overflow check so that it reverts in cases where truncation would happen.

Reviewer 2: Overflow check added.

``self`` variable. The only situation where no copy will be performed
is when storage reference variables are used or when internal library
functions are called.

Another example shows how to define a custom operator for a user-defined type:

.. code-block:: solidity

    // SPDX-License-Identifier: GPL-3.0
    pragma solidity ^0.8.18;

    type UFixed16x2 is uint16;

    using {
        add as +,
        div as /
    } for UFixed16x2;

    uint32 constant SCALE = 100;

    function add(UFixed16x2 a, UFixed16x2 b) pure returns (UFixed16x2) {
        return UFixed16x2.wrap(UFixed16x2.unwrap(a) + UFixed16x2.unwrap(b));
    }

    function div(UFixed16x2 a, UFixed16x2 b) pure returns (UFixed16x2) {
        uint32 a32 = UFixed16x2.unwrap(a);
        uint32 b32 = UFixed16x2.unwrap(b);
        uint32 result32 = a32 * SCALE / b32;

>         require(result32 <= 65536, "Divide overflow");

Reviewer: `65536` won't fit in `uint16`. But anyway, we should use `type().max` here:
```suggestion
        require(result32 < type(uint16).max, "Divide overflow");
```

Reviewer 2: fixed
